Documenting saga diagnostics (#5654)
Particular/docs.particular.net: nservicebus/sagas/analyzers.md

---
title: Roslyn analyzers for sagas
summary: Details of the Roslyn analyzers that promote code quality in sagas.
component: Core
versions: '[7.7,)'
reviewed: 2022-03-04
---

> Starting in NServiceBus version 7.7, [Roslyn analyzers](https://docs.microsoft.com/en-us/visualstudio/code-quality/roslyn-analyzers-overview) are packaged with the NServiceBus package that analyze the code in sagas and make suggestions for improvements, right in the editor.

Reviewer: ```suggestion
Starting in NServiceBus version 7.7, [Roslyn analyzers](https://docs.microsoft.com/en-us/visualstudio/code-quality/roslyn-analyzers-overview) are packaged with the NServiceBus package that analyzes the code in sagas and makes suggestions for improvements, right in the editor.
```

Author: Nope, Roslyn analyzers here is plural, so noun-verb agreement means "analyze" and "make".

* **Example message**: A saga cannot map to the saga data's Id property, regardless of casing. Select a different property (such as OrderId, CustomerId) that relates all of the messages handled by this saga.

The `Id` property of a saga (defined by the required `IContainSagaData` interface) is reserved for use by the saga. It cannot be used as a correlation id by mapping to it in the `ConfigureHowToFindSaga()` method.

In addition, some saga persistence libraries, such as [SQL Persistence](/persistence/sql/), store the saga's `Id` value in a column, and the column names are commonly case-insenstiive. This means that other casings of `Id` (`ID`, `id`, or even `iD`) are also not allowed.

## Message types should not be used as saga data properties

* **Rule ID**: NSB0010
* **Severity**: Warning
* **Example message**: Using the message type `MyMessage` to store message contents in saga data is not recommended, as it creates unnecessary coupling between the structure of the message and the stored saga data, making both more difficult to evolve.

When a saga receives a message, it can be convenient and even tempting to insert the whole thing into the saga data.

> However, this creates unintended couling between the saga data and the message contract.

Reviewer: ```suggestion
However, this creates unintended coupling between the saga data and the message contract.
```

Prior to NServiceBus version 7.7, this check was a runtime error. In NServiceBus version 7.7 and above, the analyzer diagnostic raises this feedback to compile time.

## Saga data classes should inherit ContainSagaData

* **Rule ID**: NSB0012
* **Severity**: Warning
* **Example message**: It's easier to inherit the class ContainSagaData, which contains all the necessary properties to implement IContainSagaData, than to implement IContainSagaData directly.

The generic class constraints on `Saga<TSagaData>` require the saga data class to implement the `IContainSagaData` interface, which specifies properties required by the saga infrastructure. However, it is much easier to directly inherit `ContainSagaData`, which already specifies these properties.

A benefit to inheriting the `ContainSagaData` class is that in NServiceBus version 7 and above the implemented properties are decorated with `[EditorBrowsable(EditorBrowsableState.Never)]`, which means that those properties that are _only_ needed by the saga infrastructure will not appear in IntelliSense. This makes it less likely that one of these reserved properties will be used accidentally.

> One exception comes when [using NHibernate's `[RowVersion]` attribute to control optimistic concurrency ](/persistence/nhibernate/saga-concurrency.md#custom-behavior-explicit-version). This attribute is not compatible with derived classes, so in this case the saga data class must implement `IContainSagaData` directly. In this case the `NSB0012` diagnostic can be suppressed to remove the warning.

Reviewer: ```suggestion
One exception comes when [using NHibernate's `[RowVersion]` attribute to control optimistic concurrency](/persistence/nhibernate/saga-concurrency.md#custom-behavior-explicit-version). This attribute is not compatible with derived classes, so in this case the saga data class must implement `IContainSagaData` directly. In this case the `NSB0012` diagnostic can be suppressed to remove the warning.
```
